### PR TITLE
coprocessor/codec: add comments and inline

### PR DESCRIPTION
**Pull request by siddontang**
## What have you changed? (mandatory)

Add comments and inline flag.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

No

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Add a few positive/negative examples (optional)

PTAL @breeswish @AndreMouche

### CODE REVIEW

File: src/coprocessor/codec/overflow.rs [hunk 1]
```diff
@@ -30,7 +30,7 @@ pub fn div_i64(a: i64, b: i64) -> Result<i64> {
     }
 }
 
-// div_u64_with_i64 divides u64 a with i64 b, returns:
+// `div_u64_with_i64` divides u64 a with i64 b, returns:
```
Comment (breeswish) [inline src/coprocessor/codec/overflow.rs]: ///

File: src/coprocessor/codec/overflow.rs [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 use coprocessor::codec::{Error, Result};
 
-// div_i64 divides i64 a with b, returns:
+// `div_i64` divides i64 a with b, returns:
```
Comment (breeswish) [inline src/coprocessor/codec/overflow.rs]: ///

File: src/coprocessor/codec/overflow.rs [hunk 3]
```diff
@@ -52,7 +52,7 @@ pub fn div_u64_with_i64(a: u64, b: i64) -> Result<u64> {
     }
 }
 
-// div_i64_with_u64 divides i64 a with u64 b, returns:
+// `div_i64_with_u64` divides i64 a with u64 b, returns:
```
Comment (breeswish) [inline src/coprocessor/codec/overflow.rs]: ///

File: src/coprocessor/codec/datum.rs [hunk 4]
```diff
@@ -935,21 +947,27 @@ pub fn approximate_size(values: &[Datum], comparable: bool) -> usize {
...
     encode(values, true)
 }
 
+/// `encode_value` encodes a datum slice into a buffer.
 pub fn encode_value(values: &[Datum]) -> Result<Vec<u8>> {
     encode(values, false)
 }
 
+/// `encode_to` encodes a datum slice and appends the buffer to a vector.
+/// Uses comparable to encode a memory comparable buffer or not.
```
Comment (CaitinChen) [inline src/coprocessor/codec/datum.rs]: /// Uses `comparable` to encode or not to encode a memory comparable buffer.

File: src/coprocessor/codec/table.rs [hunk 5]
```diff
@@ -81,6 +83,7 @@ pub fn check_table_ranges(ranges: &[KeyRange]) -> Result<()> {
     Ok(())
 }
 
+/// Decode table ID from the key.
```
[the remark this answers is not in the document]
Comment (CaitinChen) [reply src/coprocessor/codec/table.rs]: Yes. "Decode" -> "Decodes"

File: src/coprocessor/codec/table.rs [hunk 6]
```diff
@@ -313,29 +320,35 @@ impl RowColsDict {
         RowColsDict { value, cols }
     }
 
+    /// Returns the total count of the columns.
+    #[inline]
     pub fn len(&self) -> usize {
         self.cols.len()
     }
 
+    /// Returns whether has columns or not.
```
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: -> Returns whether columns exist or not.
or
-> Returns whether it has columns or not.

File: src/coprocessor/codec/table.rs [hunk 7]
```diff
@@ -313,29 +320,35 @@ impl RowColsDict {
...
+    #[inline]
     pub fn is_empty(&self) -> bool {
         self.cols.is_empty()
     }
 
+    /// Gets a column data from its meta if key exists.
```
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: -> Gets the column data from its meta if a key exists.
or
-> Gets the column data from its meta if `key` exists.
??

File: src/coprocessor/codec/table.rs [hunk 8]
```diff
@@ -313,29 +320,35 @@ impl RowColsDict {
...
+    /// Appends a column to the row.
     pub fn append(&mut self, cid: i64, value: &mut Vec<u8>) {
         let offset = self.value.len();
         let length = value.len();
         self.value.append(value);
         self.cols.insert(cid, RowColMeta::new(offset, length));
     }
 
-    // get binary of cols, keep the origin order, return one slice and cols' end offsets.
+    // Gets binary of cols, keep the origin order, return one slice and cols' end offsets.
```
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: /// Gets binary of cols, keeps the original order, and returns one slice and cols' end offsets.

File: src/coprocessor/codec/table.rs [hunk 9]
```diff
@@ -355,7 +368,7 @@ impl RowColsDict {
     }
 }
 
-// `cut_row` cut encoded row into (col_id,offset,length)
+// `cut_row` cuts encoded row into (col_id,offset,length)
```
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: /// `cut_row` cuts the encoded row into (col_id,offset,length)

File: src/coprocessor/codec/table.rs [hunk 10]
```diff
@@ -355,7 +368,7 @@ impl RowColsDict {
...
 // and return interested columns' meta in RowColsDict
```
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: /// and returns interested columns' meta in RowColsDict

File: src/coprocessor/codec/overflow.rs (hunk 2 above)
Comment (CaitinChen) [reply src/coprocessor/codec/overflow.rs]: /// `div_i64` divides i64 a with b and returns:

File: src/coprocessor/codec/overflow.rs (hunk 1 above)
Comment (CaitinChen) [reply src/coprocessor/codec/overflow.rs]: /// `div_u64_with_i64` divides u64 a with i64 b and returns:

File: src/coprocessor/codec/overflow.rs (hunk 3 above)
Comment (CaitinChen) [reply src/coprocessor/codec/overflow.rs]: /// `div_i64_with_u64` divides i64 a with u64 b and returns:

File: src/coprocessor/codec/table.rs (hunk 8 above)
Comment (CaitinChen) [inline src/coprocessor/codec/table.rs]: /// Gets binary of cols, keeps the original order, and returns one slice and cols' end offsets.

File: src/coprocessor/codec/datum.rs [hunk 11]
```diff
@@ -82,12 +83,16 @@ impl Debug for Datum {
...
+/// `cmp_f64` compares the f64 values and returns the Ordering.
 #[inline]
 pub fn cmp_f64(l: f64, r: f64) -> Result<Ordering> {
     l.partial_cmp(&r)
         .ok_or_else(|| invalid_type!("{} and {} can't be compared", l, r))
 }
 
+/// `checked_add_i64` checks whether `r` is larger or equal to 0.
+/// If yes, it adds `r` to the `l`, otherwise, it gets the negated version
+/// of `r` and reduces it from `l`.
```
Comment (AndreMouche) [inline src/coprocessor/codec/datum.rs]: Could we make it easier to understand like:
 `checked_add_i64` checks and adds `r` to the `l`. Return None if the sum is negative.

**Comment by siddontang** [comment]
PTAL @breeswish @AndreMouche

**Pull request merged**